tensorcircuit/mpscircuit.py

```diff
@@ -225,35 +298,224 @@ def apply_double_gate(
-                self.apply_adjacent_double_gate(
-                    gates.swap(), index, index + 1, center_position=index  # type: ignore
-                )
+            self.consecutive_swap(index2 - 1, index1)
         else:
-            self.position(index2)
-            for index in np.arange(index1 + 1, index2)[::-1]:
-                self.apply_adjacent_double_gate(
-                    gates.swap(), index, index + 1, center_position=index  # type: ignore
-                )
+            self.consecutive_swap(index2, index1 + 1)
             self.apply_adjacent_double_gate(
                 gate, index1, index1 + 1, center_position=index1 + 1
             )
-            for index in np.arange(index1 + 1, index2):
-                self.apply_adjacent_double_gate(
-                    gates.swap(), index, index + 1, center_position=index + 1  # type: ignore
-                )
+            self.consecutive_swap(index1 + 1, index2)
+
+    @classmethod
+    def gate_to_MPO(
+        cls,
+        gate: Union[Gate, Tensor],
+        *index: int,
+    ) -> Tuple[Sequence[Tensor], int]:
+        # should I put this function here?
+        """
+        Convert gate to MPO form with identities at empty sites
+        """
+        # If sites are not adjacent, insert identities in the middle, i.e.
+        #   |       |             |   |   |
+        # --A---x---B--   ->    --A---I---B--
+        #   |       |             |   |   |
+        # where
+        #      a
+        #      |
+        # --i--I--j-- = \delta_{i,j} \delta_{a,b}
+        #      |
+        #      b
+        index_left = np.min(index)
+        if isinstance(gate, tn.Node):
+            gate = backend.copy(gate.tensor)
+        index = np.array(index) - index_left
+        nindex = len(index)
+        # transform gate from (in1, in2, ..., out1, out2 ...) to
+        order = tuple(np.arange(2 * nindex).reshape((2, nindex)).T.flatten())
+        shape = (4,) * nindex
+        gate = backend.reshape(backend.transpose(gate, order), shape)
+        # (in1, out1, in2, out2, ...)
+        argsort = np.argsort(index)
+        # reorder the gate according to the site positions
+        gate = backend.transpose(gate, tuple(argsort))
+        index = index[argsort]  # type: ignore
+        # split the gate into tensors assuming they are adjacent
+        main_tensors = cls.wavefunction_to_tensors(gate, dim_phys=4, norm=False)
+        # each tensor is in shape of (i, a, b, j)
+        tensors = []
+        previous_i = None
+        for i, main_tensor in zip(index, main_tensors):
+            # insert identites in the middle
+            if previous_i is not None:
+                for _ in range(previous_i + 1, i):
+                    bond_dim = tensors[-1].shape[-1]
+                    I = (
+                        np.eye(bond_dim * 2)
+                        .reshape((bond_dim, 2, bond_dim, 2))
+                        .transpose((0, 1, 3, 2))
+                        .astype(dtypestr)
```

Comment:
why this work when dtypestr is actually `"comple64"`? numpy can recognize this?

Reply:
I think so? Numpy does have the 'complex64' dtype.